ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.cache
+
+import com.daml.caching.Cache
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.metrics.Metrics
+import com.daml.platform.store.cache.StateCache.PendingUpdates
+import com.daml.scalautil.Statement.discard
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future}
+
+/** This class is a wrapper around a Caffeine cache designed to handle
+  * correct resolution of concurrent updates for the same key.
+  */
+private[platform] case class StateCache[K, V](cache: Cache[K, V], metrics: Metrics)(implicit
+    ec: ExecutionContext
+) {
+  private val logger: ContextualizedLogger = ContextualizedLogger.get(getClass)
+  private[store] val pendingUpdates = mutable.Map.empty[K, PendingUpdates]
+
+  /** Fetch the corresponding value for an input key, if present.
+    *
+    * @param key the key to query for
+    * @return optionally [[V]]
+    */
+  def get(key: K)(implicit loggingContext: LoggingContext): Option[V] =
+    cache.getIfPresent(key) match {
+      case Some(value) =>
+        logger.debug(s"Cache hit for $key -> ${value.toString.take(100)}")
+        Some(value)
+      case None =>
+        logger.debug(s"Cache miss for $key ")
+        None
+    }
+
+  /** Update the cache synchronously.
+    *
+    * In face of multiple in-flight updates competing for the `key` (see [[putAsync()]]),
+    * this method updates the cache only if the to-be-inserted tuple is the most recent
+    * (i.e. it has `validAt` highest amongst the competing updates).
+    *
+    * @param key the key at which to update the cache
+    * @param validAt ordering discriminator for pending updates for the same key
+    * @param value the value to insert
+    */
+  def put(key: K, validAt: Long, value: V): Unit = pendingUpdates.synchronized {
+    val competingLatestForKey =
+      pendingUpdates
+        .get(key)
+        .map { pendingUpdate =>
+          val oldLatestValidAt = pendingUpdate.latestValidAt
+          pendingUpdate.latestValidAt =
+            if (pendingUpdate.latestValidAt < validAt) validAt else pendingUpdate.latestValidAt
```

Comment:
```suggestion
          pendingUpdate.latestValidAt = max(oldLatestValidAt,  validAt)
```
Using https://www.scala-lang.org/api/2.12.0/scala/math/index.html#max(x:Long,y:Long):Long

Reply:
Ah, the simplest thing. Thanks